File: client/scripts/src/api/arcgis.js
define(["jquery", "config", "collection/regions"], function($, config, regions) {
    // URL endpoint for geocoding
    var ADDRESS_URL = "https://geocode.arcgis.com/arcgis/rest/services/World/GeocodeServer/geocodeAddresses";


    var ACCESS_TOKEN = null,
        // Timestamp
        ACCESS_EXPIRATION = null;

    return {
        getStoredToken: function() { return ACCESS_TOKEN; },
        getStoredExpiration: function() { return ACCESS_EXPIRATION; },
        getAuthToken: function() {
            if (ACCESS_TOKEN && (new Date().valueOf() < ACCESS_EXPIRATION)) {
                // Pass along the existing access token
                return $.Deferred().resolve(ACCESS_TOKEN);
            }

            return $.getJSON("https://www.arcgis.com/sharing/oauth2/token", {
                grant_type: "client_credentials",
                client_id: config.clientId,
                client_secret: config.clientSecret,
                f: "pjson"
            }).then(function(json) {
                ACCESS_TOKEN = json.access_token;
                ACCESS_EXPIRATION = (new Date() + json.expires_in*1000);
                return json.access_token;
            });
        },

        geocode: function(addr) {
            return this.getAuthToken().then(function(token) {
                var activeRegions = regions.getSelection(),
                    region = activeRegions.length && activeRegions[0].attributes;
                return $.getJSON(ADDRESS_URL,
                                 {
                                     addresses: JSON.stringify({
                                         records: [
                                             {
                                                 attributes: {
                                                     OBJECTID: 1,
                                                     Address: addr,
                                                     City: region.city,
                                                     Region: region.state
                                                 }
                                             }
                                         ]
                                     }),
                                     token: token,
                                     f: "pjson"
                                 }).then(function(json) {
                                     var error;
                                     if (!json.locations ||
                                         json.locations.length === 0) {
                                         error = "No matching locations.";
                                     } else {
                                         var attrs = json.locations[0].attributes;

                                         if (!attrs.StAddr || attrs.score < 80) {
                                             error = "No good matches";
                                         } else {
                                             var regionMatches = regions.where({city: attrs.City,
                                                                                state: attrs.RegionAbbr,
                                                                                _selected: true});
                                             if (!regionMatches.length)
                                                 error = "No matches for this region.";
                                         }

                                     }

                                     if (error)
                                         return $.Deferred().reject({reason: error});

                                     return json;
                                 });
            });
        },

        /**
         * Given a JSON response from the ArcGIS geocoding API, extract
         * and return the latitude and longitude from the first address
         * it contains, along with a dictionary of attributes.
         *
         * @param {object} response
         * @return {Array} array of [lat, long, attributes]
         */
        getLatLngForFirstAddress: function(response) {
            var loc = response.locations[0].attributes;

            return [loc.Y, loc.X, loc];
        },

        reverseGeocode: function(lat, long) {
            return this.getAuthToken().then(function(token) {
                return $.getJSON("https://geocode.arcgis.com/arcgis/rest/services/World/GeocodeServer/reverseGeocode",
                      {
                          location: (long + "," + lat),
                          token: token,
                          f: "pjson"
                      });
            });
        }
    };
});

File: server/scripts/arcgis.py
import json
import re
from urllib.parse import urlencode
from urllib.request import urlopen

import requests


ADDRESS_URL = "http://geocode.arcgis.com/arcgis/rest/services/World/GeocodeServer/geocodeAddresses"
REVERSE_URL = "http://geocode.arcgis.com/arcgis/rest/services/World/GeocodeServer/reverseGeocode"


def camel_to_under(s):
    if not s:
        return ""
    return re.sub(r"(?!^)[A-Z]",
                  lambda r: ("_"+r.group(0).lower()),
                  s).lower()


def simplify(result):
    loc = result["location"]
    attrs = result["attributes"]
    return {
        "location": {
            "lng": loc["x"],
            "lat": loc["y"]
        },
        "formatted_name": attrs["Match_addr"],
        "properties": {
            "types": [camel_to_under(attrs["Addr_type"])],
            "score": result["score"]
        }
    }

def reverse_geocode(token, lat, lng):
    response = requests.get(REVERSE_URL, {"token": token,
                                          "location": f"{lng},{lat}"})
    return response.json()


class ArcGISCoder(object):
    def __init__(self, client_id, client_secret, url=ADDRESS_URL):
        assert client_id and client_secret, \
            "You must supply a client id and secret to ArcGISCoder"

        self.client_id = client_id
        self.client_secret = client_secret
        self.url = url
        self.access_token = None

    def get_access_token(self):
        if self.access_token:
            return self.access_token

        data = urlencode([("grant_type", "client_credentials"),
                          ("client_id", self.client_id),
                          ("client_secret", self.client_secret)]).\
                          encode("ISO-8859-1")
        f = urlopen("https://www.arcgis.com/sharing/oauth2/token",
                    data)
        json_response = json.loads(f.read().decode("utf-8"))

        self.access_token = json_response["access_token"]
        return self.access_token

    def geocode(self, addrs, **kwargs):
        if isinstance(addrs, str):
            addrs = [addrs]

        addresses = json.dumps(
            {"records":
             [{"attributes":
               {
                   "OBJECTID": i+1,
                   "Address": addr,
                   "City": "Somerville",
                   "Region": "MA"
               }} for i, addr in enumerate(addrs)]})
        data = {
            "addresses": addresses,
            "token": self.get_access_token(),
            "f": "json"
        }

        f = urlopen(self.url, urlencode(data).encode("ISO-8859-1"))
        result = json.loads(f.read().decode("utf-8"))

        return [simplify(l) for l in result["locations"]]

    def reverse_geocode(self, lat, lng):
        result = reverse_geocode(self.get_access_token(), lat, lng)
        return result["address"]["Address"]
